File: java/src/ggz/client/core/Game.java
/*
 * Copyright (C) 2006  Helg Bredow
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */
package ggz.client.core;

import ggz.client.mod.ModGGZ;
import ggz.client.mod.ModState;
import ggz.client.mod.ModTransactionHandler;
import ggz.client.mod.Seat;
import ggz.client.mod.SpectatorSeat;
import ggz.common.ChatType;
import ggz.common.SeatType;

import java.io.IOException;
import java.net.Socket;
import java.util.List;
import java.util.logging.Logger;

/*
 * The Game object manages information about a particular running game
 */
public class Game implements ModTransactionHandler {
    private static final Logger log = Logger.getLogger(Game.class.getName());

    /* Pointer to module this game is playing */
    private Module module;

    /* Room events */
    // HookList *event_hooks[sizeof(events) /
    // sizeof(events[0])];
    private GameEventListener event_hooks;

    /* Game module connection */
    private ModGGZ client;

    /* The server data for this game. */
    private Server server;

    /* Are we playing or watching at our table? */
    private boolean spectating;

    /* What's our seat number? */
    private int my_seat_num;

    /* Which room this game is in. */
    private int room_id;

    /* The table ID for this game. */
    private int table_id;

    /* Publicly exported functions */

    public Game(Server server, Module module) {
        if (server == null) {
            throw new IllegalArgumentException("server cannot be null");
        }
        if (server.get_cur_room() == null) {
            throw new IllegalStateException(
                    "server does not have a current room");
        }
        if (server.get_cur_game() != null) {
            throw new IllegalStateException("server already has a current game");
        }

        if (module == null)
            throw new IllegalArgumentException("module is null");

        Room room = server.get_cur_room();

        this.server = server;
        this.room_id = room.get_id();
        this.table_id = -1;

        server.set_cur_game(this);

        this.module = module;

        log.fine("Initializing new game");

        // Setup client module connection.
        this.client = new ModGGZ();
        this.client.set_server_host(server.get_host(), server.get_port(),
                server.get_handle());
        this.client.set_transaction_handler(this);
        this.client.set_player(server.get_handle(), false, -1);

        if (module.isEmbedded()) {
            this.client.set_module(module.getClassName());
        } else {
            this.client.set_module(module.getPWD(), module.getCommandLine());
        }
    }

    /* Functions for attaching hooks to Game events */
    public void add_event_hook(GameEventListener listener)
    // const GameEvent event,
    // const HookFunc func)
    {
        if (this.event_hooks != null && listener != this.event_hooks) {
            throw new UnsupportedOperationException(
                    "Multiple listeners not supported yet.");
        }
        this.event_hooks = listener;
    }

    /* Functions for removing hooks from Game events */
    public void remove_event_hook(GameEventListener l) {
        if (this.event_hooks == l) {
            this.event_hooks = null;
        }
    }

    /**
     * This function is called by ggzmod when the game state is changed.
     * 
     * Game state changes are all initiated by the game client through ggzmod.
     * So if we get here we just have to update ggzcore and the ggz client based
     * on what changes have already happened.
     */
    public void handle_state(ModState prev) throws IOException {
        ModState newState = this.client.get_state();

        log.fine("Game changing from state " + prev + " to " + newState);

        if (prev == ModState.GGZMOD_STATE_CREATED) {
            if (newState != ModState.GGZMOD_STATE_CONNECTED) {
                log.severe("Game changed state from created to " + newState);
                throw new IllegalStateException(
                        "New state != GGZMOD_STATE_CONNECTED");
            }
            log.fine("game negotiated");
            event(GameEvent.GGZ_GAME_NEGOTIATED, null);
            // Now that the game client is ready, create the channel for it. We
            // might want to consider removing this and always having game
            // clients create the channel.
            if (this.module.isEmbedded()) {
                this.server.create_channel();
            }
        } else if (prev == ModState.GGZMOD_STATE_CONNECTED) {
            if (newState != ModState.GGZMOD_STATE_WAITING
                    && newState != ModState.GGZMOD_STATE_PLAYING) {
                log.fine("Game client set state to DONE before we even began.");
                // log.severe("Game changed state from connected to " +
                // newState);
                // throw new IllegalStateException(
                // "New state != GGZMOD_STATE_WAITING or GGZMOD_STATE_PLAYING");
            } else {
                log.fine("game playing");
                event(GameEvent.GGZ_GAME_PLAYING, null);
            }
        }

        if (newState == ModState.GGZMOD_STATE_DONE) {
            Room room = this.server.get_cur_room();
            if (room != null) {
                if (this.server.get_state() == StateID.GGZ_STATE_AT_TABLE) {
                    room.leave_table(true);
                } else {
                    // Game client is exiting before we have finished the whole
                    // setup. The core client will handle this gracedully since
                    // the game will be null.
                    abort_game();
                }
            }
        } else if (newState == ModState.GGZMOD_STATE_CREATED) {
            // Leave the game running. This should never happen since this is
            // the initial state and we never return to it after leaving it.
            log.severe("Game state changed to created!");
            throw new IllegalStateException(
                    "Game state changed to created. This should never happen!");
        }
    }

    public void handle_error(Throwable exception) {
        exception.printStackTrace();
    }

    public void handle_sit(int seat_num) throws IOException {
        Net net = this.server.get_net();
        ReseatType op;

        if (this.spectating) {
            op = ReseatType.GGZ_RESEAT_SIT;
        } else {
            op = ReseatType.GGZ_RESEAT_MOVE;
        }

        net.send_table_reseat(op, seat_num);
    }

    public void handle_stand() throws IOException {
        Net net = this.server.get_net();

        net.send_table_reseat(ReseatType.GGZ_RESEAT_STAND, -1);
    }

    public void handle_boot(String name) throws IOException {
        Net net = this.server.get_net();
        Room room = this.server.get_room_by_id(this.room_id);
        Table table = room.get_table_by_id(this.table_id);
        int i;

        for (i = 0; i < table.get_num_seats(); i++) {
            TableSeat seat = table.get_nth_seat(i);

            if (seat.type != SeatType.GGZ_SEAT_PLAYER
                    || !seat.name.equals(name))
                continue;
            net.send_table_boot_update(table, seat);
            return;
        }

        for (i = 0; i < table.get_num_spectator_seats(); i++) {
            TableSeat spectator = table.get_nth_spectator_seat(i);

            if (!spectator.name.equals(name))
                continue;
            net.send_table_boot_update(table, spectator);
            return;
        }

        /*
         * If the player has already left, we won't find them and we'll end up
         * down here.
         */
    }

    public void handle_open(int seat_num) throws IOException {
        Net net = this.server.get_net();
        TableSeat seat = new TableSeat(seat_num, null, null);
        Room room = this.server.get_room_by_id(this.room_id);
        Table table = room.get_table_by_id(this.table_id);
        seat.type = SeatType.GGZ_SEAT_OPEN;
        net.send_table_seat_update(table, seat);
    }

    public void handle_bot(int seat_num) throws IOException {
        Net net = this.server.get_net();
        TableSeat seat = new TableSeat(seat_num, null, null);
        Room room = this.server.get_room_by_id(this.room_id);
        Table table = room.get_table_by_id(this.table_id);
        seat.type = SeatType.GGZ_SEAT_BOT;
        net.send_table_seat_update(table, seat);
    }

    public void handle_chat(ChatType type, String recipient, String chat)
            throws IOException {
        Room room = this.server.get_cur_room();

        room.chat(type, recipient, chat);
    }

    public void handle_info(int seat_num) throws IOException {
        Net net = this.server.get_net();

        net.send_player_info(seat_num);
    }

    void set_table(int room_id, int table_id) {
        Room room;
        Table table;
        int num_seats, i;

        room = server.get_cur_room();

        this.table_id = table_id;
        table = room.get_table_by_id(table_id);

        num_seats = table.get_num_seats();
        for (i = 0; i < num_seats; i++) {
            TableSeat seat = table.get_nth_seat(i);

            set_seat(seat);
        }

        num_seats = table.get_num_spectator_seats();
        for (i = 0; i < num_seats; i++) {
            TableSeat seat = table.get_nth_spectator_seat(i);

            if (seat.name != null) {
                set_spectator_seat(seat);
            }
        }
    }

    void set_seat(TableSeat seat) {
        Seat mseat = new Seat(seat.index, seat.type, seat.name);
        this.client.set_seat(mseat);
    }

    void set_spectator_seat(TableSeat seat) {
        SpectatorSeat mseat = new SpectatorSeat(seat.index, seat.name);
        this.client.set_spectator_seat(mseat);
    }

    void set_player(boolean is_spectator, int seat_num) {
        if (this.spectating == is_spectator && this.my_seat_num == seat_num)
            return;

        this.spectating = is_spectator;
        this.my_seat_num = seat_num;

        this.client.set_player(server.get_handle(), is_spectator, seat_num);
    }

    void set_info(List infos) {
        this.client.set_info(infos);
    }

    void inform_chat(String player, String msg) {
        this.client.inform_chat(player, msg);
    }

    /**
     * Informs the game client that an error occurred as the result of an
     * action, like "reseat".
     * 
     * @param msg
     */
    void inform_error(String msg) {
        this.client.inform_error(msg);
    }

    public boolean is_spectator() {
        return this.spectating;
    }

    public int get_seat_num() {
        return this.my_seat_num;
    }

    public int get_room_id() {
        return this.room_id;
    }

    public int get_table_id() {
        return this.table_id;
    }

    /**
     * Called after we have left the table.
     */
    public void abort_game() {
        try {
            this.client.disconnect();
        } catch (IOException ex) {
            ex.printStackTrace();
        }

        // if (room != null)
        // room.table_event(RoomEvent.GGZ_TABLE_LEFT, new TableLeaveEventData(
        // LeaveType.GGZ_LEAVE_NORMAL, null));

        server.set_cur_game(null);
    }

    // public void disconnect() throws IOException{
    // client.disconnect();
    // }

    public void set_server_fd(Socket fd) {
        this.client.set_server_fd(fd);
    }

    public Module get_module() {
        return this.module;
    }

    private void event(GameEvent event, Object data) {
        if (event_hooks == null)
            return;

        if (event == GameEvent.GGZ_GAME_LAUNCH_FAIL) {
            event_hooks.game_launch_fail((Exception) data);
        } else if (event == GameEvent.GGZ_GAME_LAUNCHED) {
            event_hooks.game_launched();
        } else if (event == GameEvent.GGZ_GAME_NEGOTIATE_FAIL) {
            event_hooks.game_negotiate_fail();
        } else if (event == GameEvent.GGZ_GAME_NEGOTIATED) {
            event_hooks.game_negotiated();
        } else if (event == GameEvent.GGZ_GAME_PLAYING) {
            event_hooks.game_playing();
        }
    }

    public void launch() {
        log.fine("Launching game of " + this.module.getName());

        try {
            this.client.connect();
            log.fine("Launched game module");
            event(GameEvent.GGZ_GAME_LAUNCHED, null);
        } catch (Throwable e) {
            log.severe("Failed to connect to game module");
            this.server.set_cur_game(null);
            event(GameEvent.GGZ_GAME_LAUNCH_FAIL, e);
        }
    }

}

File: playground/ggz-wrapper/game.c
/*
 * File: game.c
 * Author: Brent Hendricks
 * Project: GGZ Text Client 
 * Date: 3/1/01
 * $Id: game.c 6490 2004-12-15 20:10:15Z josef $
 *
 * Functions for handling game events
 *
 * Copyright (C) 2000-2002 Brent Hendricks.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA
 */

#include <stdio.h>
#include <stdlib.h>

#include <ggzcore.h>

#include "game.h"
#include "loop.h"
#include "server.h"

/* Hooks for game events */
static void game_process(void);
static GGZHookReturn game_launched(GGZGameEvent, void*, void*);
static GGZHookReturn game_launch_fail(GGZGameEvent, void*, void*);
static GGZHookReturn game_negotiated(GGZGameEvent, void*, void*);
static GGZHookReturn game_negotiate_fail(GGZGameEvent, void*, void*);
static GGZHookReturn game_playing(GGZGameEvent, void *, void*);

static GGZGame *game = NULL;
static GGZGameType *gametype = NULL;
static int gameindex = -1;
static int fd = -1;
static char *user;
static int readserver = 1;

void game_init(GGZModule *module, GGZGameType *type, int index, char *nick)
{
	user = nick;
	if (game) {
		fprintf(stderr, "You're already playing a game!\n");
		return;
	}

	gametype = type;
	gameindex = index;

	game = ggzcore_game_new();
	ggzcore_game_init(game, server, module);
	game_register(game);
	ggzcore_game_launch(game);
}


void game_quit(void)
{
}


void game_destroy(void)
{
	if (game)
		ggzcore_game_free(game);
	game = NULL;
}


static void game_process(void)
{
fprintf(stderr, "*game_process*\n");
	if (game)
		ggzcore_game_read_data(game);
fprintf(stderr, "*game_process end*\n");
}


static void channel_process(void)
{
if(!readserver) return;
fprintf(stderr, "*channel_process*\n");
fprintf(stderr, "%i\n", ggzcore_game_get_control_fd(game));
	if (server) {
		int fd = ggzcore_server_get_channel(server);
		ggzcore_server_read_data(server, fd);
	}
fprintf(stderr, "*channel_process end*\n");
}


void game_channel_connected(int fd)
{
fprintf(stderr, "**channel_connected**\n");
	loop_add_fd(fd, channel_process, NULL);
}


void game_channel_ready(int fd)
{
fprintf(stderr, "**channel_ready**\n");
	ggzcore_game_set_server_fd(game, fd);
	readserver = 0;
}


void game_register(GGZGame *game)
{
	ggzcore_game_add_event_hook(game, GGZ_GAME_LAUNCHED, game_launched);
	ggzcore_game_add_event_hook(game, GGZ_GAME_LAUNCH_FAIL, game_launch_fail);
	ggzcore_game_add_event_hook(game, GGZ_GAME_NEGOTIATED, game_negotiated);
	ggzcore_game_add_event_hook(game, GGZ_GAME_NEGOTIATE_FAIL, game_negotiate_fail);
	ggzcore_game_add_event_hook(game, GGZ_GAME_PLAYING, game_playing);
}


static GGZHookReturn game_launched(GGZGameEvent id, void* event_data, 
				   void* user_data)
{
	
	fd = ggzcore_game_get_control_fd(game);
	loop_add_fd(fd, game_process, game_destroy);

	return GGZ_HOOK_OK;
}


static GGZHookReturn game_launch_fail(GGZGameEvent id, void* event_data,
				      void* user_data)
{
	fprintf(stderr, "Launch failed: %s\n", (char*)event_data);

	return GGZ_HOOK_OK;
}


static GGZHookReturn game_negotiated(GGZGameEvent id, void* event_data,
				     void* user_data)
{

	ggzcore_server_create_channel(server);

	return GGZ_HOOK_OK;
}


static GGZHookReturn game_negotiate_fail(GGZGameEvent id, void* event_data,
				      void* user_data)
{
	fprintf(stderr, "Negotiate failed: %s\n", (char*)event_data);

	return GGZ_HOOK_OK;
}


static GGZHookReturn game_playing(GGZGameEvent id, void* event_data, void* user_data)
{
	GGZRoom *room;
	GGZTable *table;


	room = ggzcore_server_get_cur_room(server);

	if(gameindex < 0) {
		table = ggzcore_table_new();
		ggzcore_table_init(table, gametype, "Fun with gaim-ggz", 2);
		if(user!=NULL)
			ggzcore_table_set_seat(table, 1, GGZ_SEAT_RESERVED, user);
		else
			ggzcore_table_set_seat(table, 1, GGZ_SEAT_OPEN, NULL);

		ggzcore_room_launch_table(room, table);
		ggzcore_table_free(table);
	}
	else {
		ggzcore_room_join_table(room, gameindex, 0);
	}

	return GGZ_HOOK_OK;
}
